### PR TITLE
Make DropInActivity More DRY

### PR DESCRIPTION
**Pull request by sshropshire**
### Summary of changes

 - Remove `BaseActivity`
 - Create `AlertPresenter`
 - Create `DropInFragment` base class
 - Add static factory methods for all DropIn fragments

 ### Checklist

 ~- [ ] Added a changelog entry~

### Authors

- @sshropshire

### CODE REVIEW

File: Drop-In/src/main/java/com/braintreepayments/api/DialogInteraction.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.braintreepayments.api;
+
+public enum DialogInteraction {
```
Comment (sarahkoop) [inline Drop-In/src/main/java/com/braintreepayments/api/DialogInteraction.java:3]: Can this be package private and the associated callback be package private?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java [hunk 2]
```diff
@@ -269,13 +315,22 @@ private void onAddCardSubmit(DropInEvent event) {
         showCardDetailsFragment(cardNumber);
     }
 
-    private boolean shouldAddFragment(String tag) {
+    private boolean fragmentAlreadyExists(String tag) {
```
Comment (sarahkoop) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java:318]: This logic is a little confusing to me - it looks like we use this method to determine if we should show a fragment or not, but this method looks like it's checking if a fragment with this tag exists, not if it's currently visible. Is that right, or does findFragmentByTag only return a fragment if it's currently visible?

File: Drop-In/src/test/java/com/braintreepayments/api/BaseActivityUnitTest.java [hunk 3]
```diff
@@ -1,106 +0,0 @@
...
-    @Test
-    public void getDropInClient_setsClientTokenPresentWhenAClientTokenIsNotPresent() {
-        Intent intent = new Intent();
-        intent.putExtra(DropInClient.EXTRA_CHECKOUT_REQUEST, new DropInRequest());
-        setup(intent);
-
-        mActivity.getDropInClient();
-
-        assertFalse(mActivity.mClientTokenPresent);
-    }
```
Comment (sarahkoop) [inline Drop-In/src/test/java/com/braintreepayments/api/BaseActivityUnitTest.java:61]: Should these be tested for DropInActivity now too?

Comment (sshropshire) [reply Drop-In/src/test/java/com/braintreepayments/api/BaseActivityUnitTest.java:61]: When I was removing `BaseActivity` I couldn't see the value in these tests. We should have enough coverage in `DropInActivityUnitTest`.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java (hunk 2 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java:318]: Yeah visibility isn't a concern here, we just don't want to re-create a fragment if one already exists and is managed by the fragment manager. Guards us against those edge cases like a user double-tapping a button; we want to make sure we don't mistakenly put two Card Details fragments on the screen at once.

File: Drop-In/src/main/java/com/braintreepayments/api/DialogInteraction.java (hunk 1 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/DialogInteraction.java]: Done!

File: Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java (hunk 2 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java]: Went ahead and reverted this to `shouldAddFragment` that does read better.

File: Drop-In/src/main/java/com/braintreepayments/api/DialogInteractionCallback.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.braintreepayments.api;
+
+public interface DialogInteractionCallback {
```
Comment (sarahkoop) [inline Drop-In/src/main/java/com/braintreepayments/api/DialogInteractionCallback.java:3]: Can this be package private too?

**Pull request merged**